pkg/core/xds/resource.go

```diff
@@ -70,47 +93,60 @@ func (s *ResourceSet) Contains(name string, resource ResourcePayload) bool {
 	return ok
 }
 
-func (s *ResourceSet) Add(resources ...*Resource) *ResourceSet {
-	for _, resource := range resources {
-		if s.Contains(resource.Name, resource.Resource) {
-			continue
+func (s *ResourceSet) Empty() bool {
+	for _, resourceMap := range s.typeToNamesIndex {
+		if len(resourceMap) != 0 {
+			return false
 		}
-		s.resources = append(s.resources, resource)
-		s.index(resource)
 	}
-	return s
+	return true
 }
 
-func (s *ResourceSet) AddSet(set ResourceSet) *ResourceSet {
-	s.Add(set.List()...)
+func (s *ResourceSet) Add(resources ...*Resource) *ResourceSet {
+	for _, resource := range resources {
+		if s.typeToNamesIndex[s.typeName(resource.Resource)] == nil {
+			s.typeToNamesIndex[s.typeName(resource.Resource)] = map[string]*Resource{}
+		}
+		s.typeToNamesIndex[s.typeName(resource.Resource)][resource.Name] = resource
+	}
 	return s
 }
 
-func (s *ResourceSet) AddNamed(namedPayloads ...NamedResourcePayload) *ResourceSet {
-	for _, namedPayload := range namedPayloads {
-		s.Add(&Resource{
-			Name:     namedPayload.GetName(),
-			Resource: namedPayload,
-		})
+func (s *ResourceSet) Remove(typ string, name string) {
+	if s.typeToNamesIndex[typ] != nil {
+		delete(s.typeToNamesIndex[typ], name)
 	}
-	return s
 }
 
-func (s *ResourceSet) typeName(resource ResourcePayload) string {
-	return proto.MessageName(resource)
+func (s *ResourceSet) Resources(typ string) map[string]*Resource {
+	return s.typeToNamesIndex[typ]
 }
 
-func (s *ResourceSet) index(resource *Resource) {
-	if s.typeToNamesIndex == nil {
-		s.typeToNamesIndex = map[string]map[string]bool{}
+func (s *ResourceSet) AddSet(set *ResourceSet) *ResourceSet {
+	if set == nil {
+		return s
 	}
-	typeName := s.typeName(resource.Resource)
-	if s.typeToNamesIndex[typeName] == nil {
-		s.typeToNamesIndex[typeName] = map[string]bool{}
+	for typ, resources := range set.typeToNamesIndex {
+		if s.typeToNamesIndex[typ] == nil {
```

Comment:
I think you can reuse `s.Add` here

Reply:
Well, not exactly. I'd have to do `s.Add(set.List()...)` which creates a lot of lists etc.